---
layout: page
permalink: /GRP/grp-PT
---

## **1. LD clumping + thresholding**

Standard approach: Clumping + thresholding (C+T)
- Shrinkage strategy: P-value threshold
- Handling Linkage Disequilibrium: Clumping

---
<br>

### **1) Basic case/control association test**

To perform a standard case/control association analysis, use the option:

```bash
$ plink --bfile test --assoc
```

which generates a file
     
     plink.assoc	

which contains the fields:
     
     CHR     Chromosome
     SNP     SNP ID
     BP      Physical position (base-pair)
     A1      Minor allele name (based on whole sample)
     F_A     Frequency of this allele in cases
     F_U     Frequency of this allele in controls
     A2      Major allele name
     CHISQ   Basic allelic test chi-square (1df)
     P       Asymptotic p-value for this test
     OR      Estimated odds ratio (for A1, i.e. A2 is reference)

To check output file:
```bash
$ vi plink.assoc
```

	CHR     SNP         BP   A1      F_A      F_U   A2        CHISQ            P           OR
	   1   sid_0          0    G   0.5098   0.4286    A        1.324       0.2499        1.387
	   1   sid_1          1    G   0.4412    0.449    A      0.01232       0.9116       0.9689
	   1   sid_2          2    G   0.4804   0.4388    A       0.3485        0.555        1.183
	   1   sid_3          3    G   0.4608   0.4286    A         0.21       0.6468        1.139
	   1   sid_4          4    G   0.4608   0.4082    A       0.5631        0.453        1.239




---
<br>

### **2) LD-based clumping**

The --clump command is used to specify one or more result files (i.e. precomputed analyses of some kind). By default, PLINK scans these files and extracts fields with the headers SNP and P. For example:

```bash
$ plink --bfile test --clump plink.assoc --clump-p1 0.1 --clump-p2 0.1 --clump-r2 0.2 --clump-kb 250
```

There are four main parameters that determine the level of clumping:

     --clump-p1        Significance threshold for index SNPs   
     --clump-p2        Secondary significance threshold for clumped SNPs
     --clump-r2        LD threshold for clumping
     --clump-kb        Physical distance threshold for clumping


which generates a file

	plink.clumped

which contains the fields:

     CHR     Chromosome code
     F       Results fileset code (1,2,...)
     SNP     SNP identifier
     BP      Physical position of SNP (base-pairs)
     TOTAL   Total number of other SNPs in clump (i.e. passing --clump-kb and --clump-r2 thresholds)
     NSIG    Number of clumped SNPs that are not significant ( p > 0.05 )
     S05     Number of clumped SNPs 0.01 < p < 0.05
     S01     Number of clumped SNPs 0.001 < p < 0.01
     S001    Number of clumped SNPs 0.0001 < p < 0.001 
     S0001   Number of clumped SNPs p < 0.0001
     SP2     List of SNPs names (and fileset code) clumped and significant at --clump-p2

To check output file: 

```bash
$ vi plink.clumped
```

	CHR    F     SNP         BP        P    TOTAL   NSIG    S05    S01   S001  S0001    SP2
	   1    1 sid_264        264   0.000364       12      9      3      0      0      0 sid_265(1),sid_267(1),sid_268(1),sid_270   
	   1    1 sid_527        527    0.00458       19      8     10      1      0      0 sid_517(1),sid_518(1),sid_519(1),sid_520
	   1    1 sid_966        966      0.006       20     12      6      2      0      0 sid_959(1),sid_960(1),sid_961(1),sid_963   
	   1    1 sid_433        433    0.00726       17     12      5      0      0      0 sid_431(1),sid_432(1),sid_434(1),sid_435


**[Tip]** To avoid line breaks in VI editor, use ```:set nowrap```

To make a snplist:

```bash
$ awk '{print $3}' plink.clumped | sed 1d > snplist
```
```bash
$ vi snplist
```
```
sid_264
sid_527
sid_966
sid_433
sid_945

(..31 lines)
```

---
<br>

### **3) SNP scoring**

The basic command to generate a score is the --score option,

```bash
$ plink --bfile test --score plink.raw
```

For plink.raw file, we need three fields: SNP ID (snplist), effect allele and beta from summary statistics (ss.txt)

```R
R

install.packages("dplyr")
library(dplyr)

ss <- read.table("ss.txt", header=TRUE)
snplist <- read.table("snplist", header=FALSE)
#head(ss)
#snplist
names(snplist)[names(snplist)=="V1"]<-c("rs")

df <- merge(ss, snplist, by="rs")

df1 <- filter(df, df$effalt < 0)
df1 <- select(df1, rs, ref, effalt)
df1$effalt <- abs(df1$effalt)
df1

df2 <- filter(df, df$effalt > 0)
df2 <- select(df2, rs, ref, effalt)
df2

out <- rbind(df1, df2)
write.table(out, file="plink.raw", sep=" ", row.names = FALSE, col.names = FALSE, quote = FALSE)

q()
```
```bash
$ vi plink.raw
```
```
sid_264 A 0.0146693
sid_366 A 0.0175326
sid_40 A 6.371976e-06
sid_414 A 0.004027207
sid_446 A 0.005107699

(..31 lines)
```

Then, running the command above (--score) would generate a file

     plink.profile

with one individual per row and the fields:

     FID     Family ID
     IID     Individual ID
     PHENO   Phenotype for that
     CNT     Number of non-missing SNPs used for scoring
     CNT2    The number of named alleles
     SCORE   Total score for that individual

To check output file: 

```bash
$ vi plink.profile
```

      FID  IID  PHENO    CNT   CNT2    SCORE
       i0   i0      1     62     28 0.00421325
       i1   i1      2     62     36 0.00512953
       i2   i2      2     62     31 0.00484148
       i3   i3      2     62     34 0.00420184
       i4   i4      2     62     37 0.00495108
       i5   i5      1     62     31 0.00459407
       i6   i6      1     62     40 0.00506683
       i7   i7      2     62     28 0.00288433
       i8   i8      2     62     31 0.00451702
       i9   i9      2     62     30 0.0044054
      i10  i10      1     62     29 0.00383463

---
<br>


### **4) Calculate Nagelkerke's R squared**

```R
R
install.packages("fmsb")
library(fmsb)
profile <- read.table("plink.profile", header=TRUE)
profile$PHENO <- ifelse(profile$PHENO==1, as.numeric("0"), as.numeric("1"))
model = glm(profile$PHENO~profile$SCORE, family = binomial())
NagelkerkeR2(model)
```
```
$N
[1] 100

$R2
[1] 0.06977417
```

---
<br>

Go to next step,  
[2. LDpred]({% post_url 2018-11-01-grp-LDpred %})

