---
layout: page
permalink: /SMCgenetics-2018/MR/mr-osmr
---

## **1. One-Sample Mendelian Randomization**

- Lecture notes : [20181117_MR_practice.pdf](/assets/images/MR/20181117_MR_practice.pdf){: target="blank" }  

- Reference : [https://www.cog-genomics.org/plink/1.9](https://www.cog-genomics.org/plink/1.9){: target="blank" }

---
<br>

### **1) Allelic scoring** (using PLINK)
---
  
-  Apply a scoring system to all your genotypes :  ```--score [filename] sum```
	
```bash
$ plink --bfile Genotype --score scorefile/HDL_score.txt sum --out OUT/Genotype.HDL
```
```bash
$ plink --bfile Genotype --score scorefile/LDL_score.txt sum --out OUT/Genotype.LDL
```
  

-  Allelic scoring results :  ```(.profile)```

```bash
$ paste OUT/Genotype.HDL.profile OUT/Genotype.LDL.profile | awk '{print $2"\t"$6"\t"$12}' > OUT/HDL_LDL.GRS
```

---
<br>

### **2) Association** (using R)

---
<br>

#### 2-1) Data preprocessing

- Load genetic risk score(GRS) data
<br>
```bash
$R
```
```R
grs <- read.table("OUT/HDL_LDL.GRS", header=T)
colnames(grs) <- c("IID", "HDL_inc_GRS", "LDL_dec_GRS")
```
- Load epi data

```R
epi<-read.table("Epi_data/Epi.txt", header=T, sep="\t")
```
- Merge GRS data & Epi data

```R
total<-merge(epi, grs, by="IID")
total$SEX <- as.factor(total$SEX)
```
<br>

#### 2-2) Regression

- Calculate a correlation
<br>

```R
cor<-glm(MI ~  AGE + SEX + HDL + LDL, data=total, family = "binomial")
summary(cor)
```
![figure](/assets/images/MR/Correlation.png){: width="80%" height="80%"}

- Estimate a causal effect

```R
mr<-glm(MI ~  AGE + SEX + HDL_inc_GRS + LDL_dec_GRS, data=total, family = "binomial")
summary(mr)
```
![figure](/assets/images/MR/Causality.png){: width="100%" height="100%"}

---
<br>

Go to next step,  
[2. Two-sample Mendelian Randomiaztion]({% post_url 2018-11-01-mr-tsmr %})
