---
layout: page
permalink: /SMCgenetics-2018/GWAS/gwas-qc
---

## **0. Data formats**

---
<br>

### **PLINK data formats**

#### a. PLINK flat files (\*.ped, \*.map)

- **\*.ped** : information of individuals and their genotypes.
	+ <U>Columns</U>  
	FID, IID, Paternal ID, Maternal ID, Sex(*1=male; 2=female; others=unknown*),  
	Affection(*-9=unknown; 1=unaffected; 2=affected*), Genotypes(*0=missing*)
```bash
$ vi gwas.ped
____________________________
:set nowrap
:q!
```

- **\*.map** : information of variants.
	+ <U>Columns</U>  
	CHR, MarkerID, Genetic distance(*0=missing*), BP
```bash
$ head gwas.map
```

<br>

#### b. PLINK binary files (\*.bed, \*.bim, \*.fam)

- **\*.bed** : BED file is encoded binary format and include information of genotypes.

- **\*.bim** : information of variants.
	+ <U>Columns</U>  
	CHR, MarkerID, Genetic distance(*0=missing*), BP, Allele1, Allele2
```bash
$ head gwas.bim
$ wc -l gwas.bim
```

- **\*.fam** : information of individuals.
	+ <U>Columns</U>  
	FID, IID, Paternal ID, Maternal ID, Sex(*1=male; 2=female; others=unknown*),  
	Affection(*-9=unknown; 1=unaffected; 2=affected*)
```bash
$ head gwas.fam
$ wc -l gwas.bim
```
Comparing with \*.ped file.
```bash
$ head gwas.ped | cut -f1-10
```

<br>

---
---
<br>
<br>

## **1. Data Quality Control**

<br>

**※ An example of Quality control pipeline**
![QCflow](/assets/images/GWAS/QC_flow.png){: width="80%" height="80%"}


---
<br>

### **1) Variant Quality Control**

---
<br>

#### a. Filtering by thresholds
- remove MAF <0.01 SNPs :  ```--maf 0.01```
- remove missingness rate ≥0.02 (call rate <0.98) SNPs :  ```--geno 0.02```
- remove HWE test P-value <1e-06 SNPs :  ```--hwe 0.000001```

```bash
$ mkdir QC    # make directory
```
```bash
$ plink --bfile gwas --maf 0.01 --geno 0.02 --hwe 0.000001 --make-bed --out QC/gwas.1
```
```bash
$ wc -l QC/gwas.1.bim
```

---
<br>
<br>

### **2) Sample Quality Control**
<div style="text-align: right"><a class="page-link" href="{% post_url 2018-11-01-gwas-qc %}-sampleqc" target="_blank">> View full version of sample QC</a></div>

---
<br>

#### a. Filtering by threshold
- remove missingness rate ≥0.05 (call rate <0.95) individuals :  ```--mind 0.05```
- generate a lists of missingness rate statistics :  ```--missing```
- generate a lists of heterozygosity rate statistics :  ```--het```

```bash
$ plink --bfile QC/gwas.1 --mind 0.05 --missing --het --make-bed --out QC/gwas.2
```

```bash
$ head QC/gwas.2.imiss
$ head QC/gwas.2.lmiss
$ head QC/gwas.2.het
```

---
<br>

#### b. Heterozygosity rate & Missingness rate of individuals

- remove outliers (e.g. >\|5σ\|)

![hetplot](/assets/images/GWAS/gwas.2.het.png){: width="60%" height="60%"}

---
<br>

#### c. Check sex

- Compare reported sex in the **\*.fam** file with estimated sex (given genotype data).
- "STATUS" column in output file (\*.sexcheck) displays "PROBLEM" or "OK" for each individual.
- male: F>0.8, Female: F<0.2

```bash
$ plink --bfile QC/gwas.2 --check-sex --out QC/gwas.2
```

```bash
$ less QC/gwas.2.sexcheck
```

---
<br>

#### d. Relatedness

- Using [KING tool](http://people.virginia.edu/~wc9c/KING/){: target="blank" }, we can estimate the relationships between samples.
- remove one of related pairs of individuals with the 2nd or closer relationship.

---
<br>

#### e. Principal Component Analysis (PCA)

- Using **smartpca** tool of [EIGENSOFT software](https://www.hsph.harvard.edu/alkes-price/software/){: target="blank" }, we can perform PCA.
- remove outliers (e.g. >\|5σ\|) or check homogeneity of genetic variation (population structure).
- We can adjust population stratification by using results of PCA; e.g. adjusting PCs in regression model as covariates.

![pcaplot](/assets/images/GWAS/gwas.pca.png){: width="60%" height="60%"}

<br>

\> Example. 1000 Genome East Asian

![pcaplot_1kg](/assets/images/GWAS/1kG.pca.EAS.12.png){: width="80%" height="80%"}

---
<br>

Let's go to next step [**2. Imputation**]({% post_url 2018-11-01-gwas-imputation %}),
or go back to [**GWAS contents list**]({% post_url 2018-11-01-gwas %}).
